### PR TITLE
Divide and Divisible

### PR DESCRIPTION
**Pull request by xuwei-k**
- https://github.com/ekmett/contravariant/blob/v1.3.2/src/Data/Functor/Contravariant/Divisible.hs
- https://youtu.be/cB8DapKQz-I?t=20m35s
- https://github.com/ekmett/contravariant/issues/18
- https://github.com/purescript/purescript-contravariant/pull/5

### CODE REVIEW

File: core/src/main/scala/scalaz/Divisible.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+/** `Divisible` is the contravariant analogue of `scalaz.Applicative`
+ *
+ * @see [[https://github.com/ekmett/contravariant/blob/v1.3.2/src/Data/Functor/Contravariant/Divisible.hs]]
+ * @see [[https://youtu.be/cB8DapKQz-I?t=20m35s ZuriHac 2015 - Discrimination is Wrong: Improving Productivity]]
+ */
+////
+trait Divisible[F[_]] extends Divide[F] { self =>
+  ////
+
+  def conquer[A]: F[A]
```
Comment (non) [inline core/src/main/scala/scalaz/Divisible.scala]: Should this take an `A` value? i.e.

```scala
def conquer[A](a: A): F[A]
```

**Comment by jedws** [comment]
this is awesome!

File: core/src/main/scala/scalaz/Divide.scala [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    divide21(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17, a18, a19, a20, a21)(identity)
+
+  def tuple22[A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18, A19, A20, A21, A22](a1: F[A1], a2: F[A2], a3: F[A3], a4: F[A4], a5: F[A5], a6: F[A6], a7: F[A7], a8: F[A8], a9: F[A9], a10: F[A10], a11: F[A11], a12: F[A12], a13: F[A13], a14: F[A14], a15: F[A15], a16: F[A16], a17: F[A17], a18: F[A18], a19: F[A19], a20: F[A20], a21: F[A21], a22: F[A22]): F[(A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18, A19, A20, A21, A22)] =
+    divide22(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16, a17, a18, a19, a20, a21, a22)(identity)
+
+  def deriving2[A1, A2, Z](f: Z => Option[(A1, A2)])(implicit a1: F[A1], a2: F[A2]): F[Z] =
+    divide2(a1, a2)(Function.unlift(f))
```
Comment (jedws) [inline core/src/main/scala/scalaz/Divide.scala]: Not so sure about turning these into partial functions, seems dangerous. OK only for irrefutable patterns…

Comment (xuwei-k) [inline core/src/main/scala/scalaz/Divide.scala]: yeah, any good suggestions?

- remove `derivingN`
- chenge `def deriving2[A1, A2, Z](f: Z => Option[(A1, A2)])`  to  `def deriving2[A1, A2, Z](f: Z => (A1, A2))`
- something else?

Comment (jedws) [inline core/src/main/scala/scalaz/Divide.scala]: I think they either need to be option 2, or be constrained by a `PlusEmpty[F]`

Comment (tonymorris) [inline core/src/main/scala/scalaz/Divide.scala]: I think second point is most appropriate.

Comment (xuwei-k) [inline core/src/main/scala/scalaz/Divide.scala]: fixed

**Pull request merged**